### ISSUE TITLE
Quarkus compatibility

### ISSUE THREAD

**Issue by solsson** [caoccao/Javet#107]
Have you investigated [Quarkus](https://quarkus.io/) compatibility? Do you think it's realistic to support native builds using GraalVM? If could do some tests if this topic is of interest.

**Comment by caoccao** [comment]
I haven't spent time evaluating if Javet is compatible with Quarkus. I heavily use Spring + Spring Boot both of which work seamlessly with Javet.

I don't see there's anything against Quarkus in Javet. I'd like to rely on the Javet community evaluating that. e.g.

* Windows 7 compatibility with the official build was confirmed by some Javet users.
* CentOS 7 compatibility with the private build was confirmed by some Javet users.

I believe Javet can work well with Quarkus, GraalVM, native image, etc. Especially, in the native image GraalJS loses the Java and JavaScript interoperability, but Javet still supports the complete interoperability. Some potential Javet users ask me whether they shall choose GraalJS or Javet. My answer is always: **Go with Javet**.

| Feature | Javet | GraalJS |
|---|:-:|:-:|
| Java and JavaScript interoperability in native image | Yes | No |
| Compatible with Chrome / V8 / Node.js | Yes |  No |
| GC | Separate (JVM + V8) | Unified |
| Performance | Best | Warm-up |

In summary, for interoperability in native image, Javet seems to be the only option in the Java world for now. If you are interested, you are welcome evaluating that. And, if you are happy with contributing back to the community, you may submit pull request with your conclusion to /docs.

Thank you and please let me know if you have any questions.

**Comment by solsson** [comment]
I found the time to learn how to write a Quarkus extension. It's promising in terms of how it can let Javet stay framework agostic while allowing applications to inject Javet `V8Host` instances (or some other entrypoint) without worrying about how to configure them for different Quarkus modes (my target is native binary container images).

Unfortunately I struggle with `setLibLoadingListerer` in Quarkus. See https://github.com/Yolean/quarkus-javet/issues/1. Could there be other ways to disable lib loading, or change the folder? Would it for example be possible to pass a custom JavetLibLoader or IJavetLibLoadingListener to `V8Host.get[Mode]Instance();`? That might be semantically wrong, because lib loading must be global. Could there be a singleton instance manager? When lib loading works, the extension will have to settle on what applications should `@Inject` and what scope it should be.

I'd be happy to donate the extension to the Javet community. Maybe a good next step is to find a devloop with Javet and extension source together. Currently everything is quite rough and I expect that anyone who isn't on Ubuntu needs to tweak the build config outlined by the readme.

**Comment by caoccao** [comment]
Thank you for raising this issue. That makes sense.

The challenge in the lib loading design is **the lib can only be loaded once regardless of how many classloaders there are**. In other words, there is only one copy of the lib in the memory. So, Javet leverages the classic singleton pattern to represent this unique memory copy of the lib. That singleton pattern comprises of 2 inner classes each of which holds a static instance of a V8 host.

This design implies there is no way of instructing the lib loader to do some custom work at instance level, because referencing the class itself triggers the execution of the static block, in other words, it's always too late. The effective way is to register the custom stuff to another static class via `JavetLibLoader#setLibLoadingListener()`.

The other way I can think of is injecting the custom stuff via environment variables. I just submitted this [commit](https://github.com/caoccao/Javet/commit/a7dc048b665166d77c532f066281282fb7cdb1de) with code and doc to the dev branch. You are welcome taking a try ([actions](https://github.com/caoccao/Javet/actions)) and please let me know if you have any questions.

**Comment by solsson** [comment]
I can confirm that the system properties have effect on native build, when passed to the graalvm compiler using `-Dquarkus.native.additional-build-args=-J-Djavet.lib.loading.type=system`. I did a [quick test](https://github.com/Yolean/quarkus-javet/commit/4decc7752cf799badc7014fd7425b17e976d208b) with `-Djavet.lib.loading.path=/tmp` but was surprised to see that a subfolder is still added.

**Comment by caoccao** [comment]
My bad. I forgot to make this [change](https://github.com/caoccao/Javet/commit/e30ff27b271701bf1be74e93ba2150e10a7e2a6c) to disable the internal auto-deployment.
